### PR TITLE
Test 120: Interchainquery module and withdrawal query + sweep 

### PR DESCRIPTION
**Pull request by riley-stride**
### Summary
Adds most updates interchainquery module, latest icq in local build subdir, makes bank balances ICQ query to withdrawal account every 3 epochs past epoch 50, then ICA bank sends 90% of the rewards to the delegation account and 10% to the stride hostzone revenue account. The ICA bank send logic is in `x/stakeibs/keeper/callbacks.go`.

#### Specific changes:
- update `x/interchainquery` to latest version
- add logic to sweep withdrawalAccount balances to delegationAcct and revAccount, in `x/stakeibc/keeper/callbacks.go`. We round protocol fees down the we round the delegation up. 
- in `x/stakeibc/keeper/msg_server_submit_tx.go`, issue the ICQ to the withdrawal acct bank balances whose callback sets that value as an attribute of hostZone proto and sweeps balances
- Add a few helpers for hooks/epochs logic, e.g. `GetLightClientHeightSafely `
- add "revenue" account on host zone in `init_gaia.sh` to hold the protocol's commission
- In `delegation.go:ProcessDelegationStaking`, code the balance to be staked to `0`, waiting for depositRecord logic

### Test plan
To replicate, set up local dev mode per Sam's PR, then:
Build:
`make local-init build=si`
Run:
`make start`
Once gaia is running, send some tokens to the withdrawal account to test the ICA bank send:
`build/gaiad --home ./scripts-local/state/gaia tx bank send gval1 cosmos1lcnmjwjy2lnqged5pnrc0cstz0r88rttunla4zxv84mee30g2q3q48fm53 3333uatom --keyring-backend test --chain-id GAIA -y`

In a different shell, to kick off the logs run:
`while true; do make logs ; sleep 3 ; done`
Watch the logs in `stride/scripts-local/logs/accounts.log` and `stride/scripts-local/logs/icq.log` and `stride/scripts-local/logs/stride.log`.

### CODE REVIEW

File: Dockerfile.icq [hunk 1]
```diff
@@ -19,8 +20,8 @@ RUN mkdir stride_ref/stride
 COPY . /src/app/stride_ref/stride
 WORKDIR /src/app/stride_ref/stride
 WORKDIR /src/app
-RUN go get github.com/Stride-Labs/interchain-queries/cmd
-RUN go get github.com/Stride-Labs/interchain-queries/pkg/runner
+# RUN go get github.com/Stride-Labs/interchain-queries/cmd
```
Comment (asalzmann) [inline Dockerfile.icq:23]: nit: rm

File: proto/stakeibc/ica_account.proto [hunk 2]
```diff
@@ -15,8 +15,7 @@ enum ICAAccountType {
 // TODO(TEST-XX): Update these fields to be more useful (e.g. balances should be coins, maybe store port name directly)
 message ICAAccount {
   string address = 1 [ (cosmos_proto.scalar) = "cosmos.AddressString" ]; 
-  int64 balance = 2; 
-  int64 delegatedBalance = 3; 
   repeated Delegation delegations = 4;
   ICAAccountType target = 5;
+  int64 withdrawalBalance = 8;
```
Comment (asalzmann) [inline proto/stakeibc/ica_account.proto:20]: add ticket to make cosmos.Dec type

Comment (asalzmann) [reply proto/stakeibc/ica_account.proto:20]: rename to `balance`

File: scripts/tests/3.sh [hunk 3]
```diff
@@ -19,13 +19,13 @@ GETSTAKE() {
 # $GAIA1_EXEC q distribution rewards $GAIA_WITHDRAWAL
 
 # echo "-----------------------------"
-
```
Comment (asalzmann) [inline scripts/tests/3.sh:22]: can we remove this file?

File: x/stakeibc/keeper/callbacks.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+func (c Callbacks) AddCallback(id string, fn interface{}) types.QueryCallbacks {
+	c.callbacks[id] = fn.(Callback)
+	return c
+}
+
+func (c Callbacks) RegisterCallbacks() types.QueryCallbacks {
+	a := c.
+		// AddCallback("valset", Callback(ValsetCallback)).
```
Comment (asalzmann) [inline x/stakeibc/keeper/callbacks.go:47]: nit: rm

File: x/stakeibc/keeper/hooks.go [hunk 5]
```diff
@@ -54,39 +54,18 @@ func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochN
...
-				}
-				// err := k.ReinvestRewards(ctx, zoneInfo)
-				// if err != nil {
-				// 	k.Logger(ctx).Error("Did not withdraw rewards on %s", zoneInfo.ChainId)
-				// 	return true
-				// } else {
-				// 	k.Logger(ctx).Info(fmt.Sprintf("Successfully withdrew rewards on %s", zoneInfo.ChainId))
-				// }
-				return false
+		if epochNumber%reinvestInterval == 0 && (epochNumber > 50) { // allow a few blocks from UpdateUndelegatedBal to avoid conflicts
```
Comment (asalzmann) [inline x/stakeibc/keeper/hooks.go:65]: can we fix this bf merging?

File: x/stakeibc/keeper/msg_server_submit_tx.go [hunk 6]
```diff
@@ -206,3 +165,24 @@ func (k Keeper) SubmitTxs(ctx sdk.Context, connectionId string, msgs []sdk.Msg,
 
 	return nil
 }
+
+func (k Keeper) GetLightClientHeightSafely(ctx sdk.Context, connectionID string) (int64, bool) {
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_submit_tx.go:169]: do we need this?

File: x/stakeibc/keeper/msg_server_submit_tx.go [hunk 7]
```diff
@@ -94,73 +94,32 @@ func (k Keeper) DelegateOnHost(ctx sdk.Context, hostZone types.HostZone, amt sdk
...
+	withdrawalIca := zoneInfo.GetWithdrawalAccount()
+	if withdrawalIca == nil || withdrawalIca.Address == "" {
+		k.Logger(ctx).Error("Zone %s is missing a delegation address!", zoneInfo.ChainId)
+	}
+	k.Logger(ctx).Info(fmt.Sprintf("\tQuerying withdrawalBalances for %s at %d height", zoneInfo.ChainId))
+
+	_, addr, _ := bech32.DecodeAndConvert(withdrawalIca.GetAddress())
+	data := bankTypes.CreateAccountBalancesPrefix(addr)
+	key := "store/bank/key"
+	k.Logger(ctx).Info("Querying for value", "key", key, "denom", zoneInfo.HostDenom) // debug?
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_submit_tx.go:110]: debug?

File: x/stakeibc/keeper/msg_server_submit_tx.go [hunk 8]
```diff
@@ -94,73 +94,32 @@ func (k Keeper) DelegateOnHost(ctx sdk.Context, hostZone types.HostZone, amt sdk
...
+	k.InterchainQueryKeeper.MakeRequest(
+		ctx,
+		zoneInfo.ConnectionId,
+		zoneInfo.ChainId,
+		key,
+		append(data, []byte(zoneInfo.HostDenom)...),
+		sdk.NewInt(-1),
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_submit_tx.go:117]: why -1?

File: x/stakeibc/keeper/msg_server_submit_tx.go [hunk 9]
```diff
@@ -94,73 +94,32 @@ func (k Keeper) DelegateOnHost(ctx sdk.Context, hostZone types.HostZone, amt sdk
...
+		types.ModuleName,
+		"withdrawalbalance",
+		0, //ttl
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_submit_tx.go:120]: why is ttl 0?

**Review by asalzmann — COMMENTED** [review]
submitting comments w/o callbacks.go review

File: x/stakeibc/keeper/hooks.go (hunk 5 above)
Comment (riley-stride) [reply x/stakeibc/keeper/hooks.go:65]: yes fixed

File: x/stakeibc/keeper/msg_server_submit_tx.go (hunk 8 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_submit_tx.go:117]: no timeout

File: x/stakeibc/keeper/msg_server_submit_tx.go (hunk 6 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_submit_tx.go:169]: unused currently, it updated the LC to latest height,  helpful to check HZ attributes

File: x/stakeibc/keeper/msg_server_submit_tx.go (hunk 9 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_submit_tx.go:120]: I believe nonzero ttl stores queries for more than one epoch rather then processing immediately. We don't need that feature

File: x/stakeibc/keeper/callbacks.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+				break
+			}
+
+		}
+		// if balance is nil, the response sent back is nil, so we don't receive the denom. Override that now.
+		coin = sdk.NewCoin(denom, sdk.ZeroInt())
+	}
+
+	// Set withdrawal balance as attribute on HostZone's withdrawal ICA account
+	wa := zone.WithdrawalAccount
```
Comment (asalzmann) [inline x/stakeibc/keeper/callbacks.go:89]: nit: GetWithdrawalAccount()

File: x/stakeibc/keeper/callbacks.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	strideClaim := strideCommission.Mul(withdrawalBalance)
+	strideClaimFloored := strideClaim.TruncateInt()
+
+	reinvestAmount := (sdk.NewDec(1).Sub(strideCommission)).Mul(withdrawalBalance)
+	reinvestAmountCeil := reinvestAmount.Ceil().TruncateInt()
+
+	// TODO(TEST-112) safety check, balances should add to original amount
+	if (strideClaimFloored.Int64() + reinvestAmountCeil.Int64()) != coin.Amount.Int64() {
+		ctx.Logger().Error(fmt.Sprintf("Error with withdraw logic: %d, Fee portion: %d, reinvestPortion %d", coin.Amount.Int64(), strideClaimFloored.Int64(), reinvestAmountCeil.Int64()))
+		panic("Failed to subdivide rewards to feeAccount and delegationAccount")
```
Comment (asalzmann) [inline x/stakeibc/keeper/callbacks.go:123]: Will this crash the chain? Don't think so bc SubmitQueryResponse is a standard tx but want to be sure

**Review by asalzmann — COMMENTED** [review]
lgtm - we should review the byte query stuff in more depth at some point

File: x/stakeibc/keeper/callbacks.go (hunk 11 above)
Comment (riley-stride) [reply x/stakeibc/keeper/callbacks.go:125]: don't think so; added todo to revisit in safety review

**Pull request merged**